File: ui/src/trace_processor/proto_ring_buffer.ts
// Copyright (C) 2021 The Android Open Source Project
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {assertTrue} from '../base/logging';

// This class is the TypeScript equivalent of the identically-named C++ class in
// //protozero/proto_ring_buffer.h. See comments in that header for a detailed
// description. The architecture is identical.

const kGrowBytes = 128 * 1024;
const kMaxMsgSize = 64 * 1024 * 1024;

export class ProtoRingBuffer {
  private buf = new Uint8Array(kGrowBytes);
  private fastpath?: Uint8Array;
  private rd = 0;
  private wr = 0;

  // The caller must call ReadMessage() after each append() call.
  // The |data| might be either copied in the internal ring buffer or returned
  // (% subarray()) to the next ReadMessage() call.
  append(data: Uint8Array) {
    assertTrue(this.wr <= this.buf.length);
    assertTrue(this.rd <= this.wr);

    // If the last call to ReadMessage() consumed all the data in the buffer and
    // there are no incomplete messages pending, restart from the beginning
    // rather than keep ringing. This is the most common case.
    if (this.rd === this.wr) {
      this.rd = this.wr = 0;
    }

    // The caller is expected to issue a ReadMessage() after each append().
    const dataLen = data.length;
    if (dataLen === 0) return;
    assertTrue(this.fastpath === undefined);
    if (this.rd === this.wr) {
      const msg = ProtoRingBuffer.tryReadMessage(data, 0, dataLen);
      if (msg !== undefined &&
          ((msg.byteOffset + msg.length) === (data.byteOffset + dataLen))) {
        // Fastpath: in many cases, the underlying stream will effectively
        // preserve the atomicity of messages for most small messages.
        // In this case we can avoid the extra buffer roundtrip and return the
        // original array (actually a subarray that skips the proto header).
        // The next call to ReadMessage() will return this.
        this.fastpath = msg;
        return;
      }
    }

    let avail = this.buf.length - this.wr;
    if (dataLen > avail) {
      // This whole section should be hit extremely rarely.

      // Try first just recompacting the buffer by moving everything to the
      // left. This can happen if we received "a message and a bit" on each
      // append() call.
      this.buf.copyWithin(0, this.rd, this.wr);
      avail += this.rd;
      this.wr -= this.rd;
      this.rd = 0;
      if (dataLen > avail) {
        // Still not enough, expand the buffer.
        let newSize = this.buf.length;
        while (dataLen > newSize - this.wr) {
          newSize += kGrowBytes;
        }
        assertTrue(newSize <= kMaxMsgSize * 2);
        const newBuf = new Uint8Array(newSize);
        newBuf.set(this.buf);
        this.buf = newBuf;
        // No need to touch rd / wr.
      }
    }

    // Append the received data at the end of the ring buffer.
    this.buf.set(data, this.wr);
    this.wr += dataLen;
  }

  // Tries to extract a message from the ring buffer. If there is no message,
  // or if the current message is still incomplete, returns undefined.
  // The caller is expected to call this in a loop until it returns undefined.
  // Note that a single write to Append() can yield more than one message
  // (see ProtoRingBufferTest.CoalescingStream in the unittest).
  readMessage(): Uint8Array|undefined {
    if (this.fastpath !== undefined) {
      assertTrue(this.rd === this.wr);
      const msg = this.fastpath;
      this.fastpath = undefined;
      return msg;
    }
    assertTrue(this.rd <= this.wr);
    if (this.rd >= this.wr) {
      return undefined;  // Completely empty.
    }
    const msg = ProtoRingBuffer.tryReadMessage(this.buf, this.rd, this.wr);
    if (msg === undefined) return undefined;
    assertTrue(msg.buffer === this.buf.buffer);
    assertTrue(this.buf.byteOffset === 0);
    this.rd = msg.byteOffset + msg.length;

    // Deliberately returning a copy of the data with slice(). In various cases
    // (streaming query response) the caller will hold onto the returned buffer.
    // If we get to this point, |msg| is a view of the circular buffer that we
    // will overwrite on the next calls to append().
    return msg.slice();
  }

  private static tryReadMessage(
      data: Uint8Array, dataStart: number, dataEnd: number): Uint8Array
      |undefined {
    assertTrue(dataEnd <= data.length);
    let pos = dataStart;
    if (pos >= dataEnd) return undefined;
    const tag = data[pos++];  // Assume one-byte tag.
    if (tag >= 0x80 || (tag & 0x07) !== 2 /* len delimited */) {
      throw new Error(
          `RPC framing error, unexpected tag ${tag} @ offset ${pos - 1}`);
    }

    let len = 0;
    for (let shift = 0; /* no check */; shift += 7) {
      if (pos >= dataEnd) {
        return undefined;  // Not enough data to read varint.
      }
      const val = data[pos++];
      len |= ((val & 0x7f) << shift) >>> 0;
      if (val < 0x80) break;
    }

    if (len >= kMaxMsgSize) {
      throw new Error(
          `RPC framing error, message too large (${len} > ${kMaxMsgSize}`);
    }
    const end = pos + len;
    if (end > dataEnd) return undefined;

    // This is a subarray() and not a slice() because in the |fastpath| case
    // we want to just return the original buffer pushed by append().
    // In the slow-path (ring-buffer) case, the readMessage() above will create
    // a copy via slice() before returning it.
    return data.subarray(pos, end);
  }
}

File: src/protozero/proto_ring_buffer.cc
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "perfetto/ext/protozero/proto_ring_buffer.h"

#include <algorithm>
#include <atomic>
#include <new>
#include <utility>

#include "perfetto/base/compiler.h"
#include "perfetto/base/logging.h"
#include "perfetto/ext/base/paged_memory.h"
#include "perfetto/protozero/proto_utils.h"

namespace protozero {

namespace {
constexpr size_t kGrowBytes = 128 * 1024;

// The boundaries of a message inside the buffer, before it is turned into a
// Message holding a reference to that buffer.
struct Token {
  const uint8_t* start = nullptr;
  uint32_t len = 0;
  uint32_t field_id = 0;
  bool fatal_framing_error = false;
  bool valid() const { return !!start; }
};

inline Token FramingError() {
  Token msg{};
  msg.fatal_framing_error = true;
  return msg;
}

// Tries to decode a length-delimited proto field from |start|.
// Returns a valid boundary if the preamble is valid and the length is within
// |end|, or an invalid message otherwise.
Token TryReadProtoMessage(const uint8_t* start, const uint8_t* end) {
  namespace proto_utils = protozero::proto_utils;
  uint64_t field_tag = 0;
  auto* start_of_len = proto_utils::ParseVarInt(start, end, &field_tag);
  if (start_of_len == start)
    return Token{};  // Not enough data.

  const uint32_t tag = field_tag & 0x07;
  if (tag !=
      static_cast<uint32_t>(proto_utils::ProtoWireType::kLengthDelimited)) {
    PERFETTO_ELOG("RPC framing error, unexpected msg tag 0x%xu", tag);
    return FramingError();
  }

  uint64_t msg_len = 0;
  auto* start_of_msg = proto_utils::ParseVarInt(start_of_len, end, &msg_len);
  if (start_of_msg == start_of_len)
    return Token{};  // Not enough data.

  if (msg_len > ProtoRingBuffer::kMaxMsgSize) {
    PERFETTO_ELOG("RPC framing error, message too large (%" PRIu64 " > %zu)",
                  msg_len, ProtoRingBuffer::kMaxMsgSize);
    return FramingError();
  }

  if (start_of_msg + msg_len > end)
    return Token{};  // Not enough data.

  Token msg{};
  msg.start = start_of_msg;
  msg.len = static_cast<uint32_t>(msg_len);
  msg.field_id = static_cast<uint32_t>(field_tag >> 3);
  return msg;
}

}  // namespace

// The memory messages are tokenized out of. Refcounted, because the messages
// handed out are slices of it and may outlive the ring buffer.
class ProtoRingBuffer::Buffer {
 public:
  static BufferPtr Create(size_t capacity) {
    return BufferPtr(new Buffer(capacity));
  }
  BufferPtr Share() {
    refs_.fetch_add(1, std::memory_order_relaxed);
    return BufferPtr(this);
  }

  // True if no Message points into this buffer, i.e. its bytes can be recycled
  // or moved around.
  bool IsUniquelyOwned() const {
    return refs_.load(std::memory_order_acquire) == 1;
  }

  uint8_t* data() { return static_cast<uint8_t*>(mem_.Get()); }
  size_t size() const { return mem_.size(); }

 private:
  friend struct BufferDeleter;

  explicit Buffer(size_t capacity)
      : mem_(perfetto::base::PagedMemory::Allocate(capacity)) {}

  void Release() {
    if (refs_.fetch_sub(1, std::memory_order_acq_rel) == 1)
      delete this;
  }

  std::atomic<uint32_t> refs_{1};
  perfetto::base::PagedMemory mem_;
};

void ProtoRingBuffer::BufferDeleter::operator()(Buffer* buffer) const {
  buffer->Release();
}

ProtoRingBuffer::Message::~Message() = default;

ProtoRingBuffer::Message::Message(Message&& other) noexcept {
  *this = std::move(other);
}

ProtoRingBuffer::Message& ProtoRingBuffer::Message::operator=(
    Message&& other) noexcept {
  if (this == &other)
    return *this;
  buf_ = std::move(other.buf_);
  start_ = std::exchange(other.start_, nullptr);
  len_ = std::exchange(other.len_, 0);
  field_id_ = std::exchange(other.field_id_, 0);
  fatal_framing_error_ = std::exchange(other.fatal_framing_error_, false);
  return *this;
}

ProtoRingBuffer::ProtoRingBuffer() : buf_(Buffer::Create(kGrowBytes)) {}
ProtoRingBuffer::~ProtoRingBuffer() = default;

size_t ProtoRingBuffer::cached_capacity_for_testing() const {
  return spare_ ? spare_->size() : 0;
}

ProtoRingBuffer::BufferPtr ProtoRingBuffer::AcquireBuffer(size_t capacity) {
  if (spare_ && spare_->IsUniquelyOwned() && spare_->size() >= capacity)
    return std::move(spare_);
  ++num_buffers_;
  return Buffer::Create(capacity);
}

void ProtoRingBuffer::RecycleBuffer(BufferPtr buffer) {
  // Keep only the most recent buffer, and only if it can satisfy future
  // replacements. In particular, growth must not cache an undersized buffer.
  spare_.reset();
  if (buffer->size() >= buf_->size())
    spare_ = std::move(buffer);
}

ProtoRingBuffer::WriteHandle::~WriteHandle() {
  PERFETTO_CHECK(buffer_ == nullptr);
}

ProtoRingBuffer::WriteHandle::WriteHandle(WriteHandle&& other) noexcept
    : buffer_(other.buffer_), data_(other.data_), size_(other.size_) {
  other.buffer_ = nullptr;
}

ProtoRingBuffer::WriteHandle& ProtoRingBuffer::WriteHandle::operator=(
    WriteHandle&& other) noexcept {
  this->~WriteHandle();  // CHECKs that any reservation held was consumed.
  new (this) WriteHandle(std::move(other));
  return *this;
}

ProtoRingBuffer::WriteHandle ProtoRingBuffer::BeginWrite(size_t data_len) {
  PERFETTO_CHECK(data_len <= kMaxMsgSize);
  // A second reservation could recompact or grow the buffer under the first.
  PERFETTO_CHECK(!write_in_flight_);
  PERFETTO_DCHECK(wr_ <= buf_->size());
  PERFETTO_DCHECK(wr_ >= rd_);

  const bool uniquely_owned = buf_->IsUniquelyOwned();
  // After a framing error, no unread bytes need preserving. Messages already
  // handed out still own their bytes, so rewinding requires exclusive
  // ownership.
  if (PERFETTO_UNLIKELY(failed_))
    rd_ = wr_;
  if (rd_ == wr_ && uniquely_owned)
    rd_ = wr_ = 0;

  if (PERFETTO_UNLIKELY(data_len > buf_->size() - wr_)) {
    size_t pending = wr_ - rd_;
    size_t required = pending + data_len;
    if (required > kMaxMsgSize * 2) {
      // These bytes can never amount to a message (e.g. a never-ending varint).
      // Reserve space for the write, which FinishWrite() will discard.
      failed_ = true;
      pending = 0;
      required = data_len;
    }

    if (uniquely_owned && required <= buf_->size()) {
      memmove(buf_->data(), buf_->data() + rd_, pending);
    } else {
      // Choose the final capacity before acquiring a buffer, so a write that
      // must preserve retained messages and grow copies unread bytes only once.
      const size_t capacity =
          std::max(buf_->size(),
                   ((required + kGrowBytes - 1) / kGrowBytes) * kGrowBytes);
      auto next = AcquireBuffer(capacity);
      memcpy(next->data(), buf_->data() + rd_, pending);
      auto previous = std::move(buf_);
      buf_ = std::move(next);
      RecycleBuffer(std::move(previous));
    }
    rd_ = 0;
    wr_ = pending;
  }

  write_in_flight_ = true;
  return WriteHandle(this, buf_->data() + wr_, data_len);
}

void ProtoRingBuffer::WriteHandle::EndWrite(size_t size_written) {
  PERFETTO_CHECK(buffer_ != nullptr);
  PERFETTO_CHECK(size_written <= size_);
  ProtoRingBuffer* buffer = buffer_;
  buffer_ = nullptr;
  buffer->FinishWrite(size_written);
}

void ProtoRingBuffer::WriteHandle::AbortWrite() {
  PERFETTO_CHECK(buffer_ != nullptr);
  ProtoRingBuffer* buffer = buffer_;
  buffer_ = nullptr;
  buffer->FinishWrite(0);
}

void ProtoRingBuffer::FinishWrite(size_t size_written) {
  write_in_flight_ = false;
  if (PERFETTO_LIKELY(!failed_))
    wr_ += size_written;
}

ProtoRingBuffer::Message ProtoRingBuffer::ReadMessage() {
  // Token only carries boundaries; attaching |buf_| is what keeps the bytes
  // alive for as long as the caller holds the Message.
  auto make_message = [this](const Token& tok) {
    Message msg;
    msg.fatal_framing_error_ = tok.fatal_framing_error;
    if (tok.valid()) {
      msg.buf_ = buf_->Share();
      msg.start_ = tok.start;
      msg.len_ = tok.len;
      msg.field_id_ = tok.field_id;
    }
    return msg;
  };

  if (failed_)
    return make_message(FramingError());

  uint8_t* buf = buf_->data();

  PERFETTO_DCHECK(rd_ <= wr_);
  if (rd_ >= wr_)
    return Message{};  // Completely empty.

  auto msg = TryReadProtoMessage(&buf[rd_], &buf[wr_]);
  if (!msg.valid()) {
    failed_ = failed_ || msg.fatal_framing_error;
    return make_message(msg);  // Could still be a framing error.
  }

  const uint8_t* msg_end = msg.start + msg.len;
  PERFETTO_CHECK(msg_end > &buf[rd_] && msg_end <= &buf[wr_]);
  auto msg_outer_len = static_cast<size_t>(msg_end - &buf[rd_]);
  rd_ += msg_outer_len;
  return make_message(msg);
}

}  // namespace protozero
